src/encoding/bencode.rs

```diff
@@ -1,136 +1,125 @@
     }
 }
 
+
 /*
 **  Integers are encoded as follows: i<integer encoded in base ten ASCII>e
 **  The initial i and trailing e are beginning and ending delimiters.
 */
-
-pub struct Int {
-    content: isize
-}
-
-impl Literal for Int {
-
-    fn stringify(&self) -> Result<String, Error> {
-        Ok(format!("i{}e", self.content))
-    }
+macro_rules! encode_int {
+    ($int: ty) => {
+        impl Encodable for $int {
+            fn encode(&self) -> Result<String, Error> {
+                Ok(format!("i{}e", self))
+            }
+        }
+    };
 }
+encode_int!(i8);    encode_int!(u8);
+encode_int!(i16);   encode_int!(u16);
+encode_int!(i32);   encode_int!(u32);
+encode_int!(i64);   encode_int!(u64);
+encode_int!(isize); encode_int!(usize);
 
-impl From<isize> for Int {
-    fn from(content: isize) -> Int {
-        Int { content }
-    }
-}
 
 /*
 **  Lists are encoded as follows: l<bencoded values>e
 **  The initial l and trailing e are beginning and ending delimiters.
 **  Lists may contain any bencoded type, including integers, strings, dictionaries, and even lists within other lists.
 */
-
-pub struct List {
-    content: Vec<ListEntry>
-}
-
-impl Literal for List {
-    fn stringify(&self) -> Result<String, Error> {
-        let mut ascii = String::new();
-        for i in self.content.iter() {
-            let result = i.stringify()?;
-            ascii.push_str(result.as_str());
+impl Encodable for Vec<Box<dyn Encodable>> {
+    fn encode(&self) -> Result<String, Error> {
+        let mut bstring: Vec<char> = Vec::new();
+        
+        for item in self {
+            bstring.extend(item.encode()?.chars())
         }
-        Ok(format!("l{}e", ascii))
-    }
-}
 
-impl From<Vec<ListEntry>> for List {
-    fn from(content: Vec<ListEntry>) -> Self {
-        List { content }
+        Ok(format!("l{}e", bstring.iter().collect::<String>()))
     }
 }
 
+
 /*
 **  Dictionaries are encoded as follows: d<bencoded string><bencoded element>e
 **  The initial d and trailing e are the beginning and ending delimiters.
 **  The values may be integers, strings, lists, and other dictionaries.
 */
-
-pub struct Dict {
-    content: Vec<DictEntry>
+// Dictionary encoding is seperate from encoding a regular vector as each entry needs to be sorted in
+// lexicogrpahical order of the key as per specification.
+pub type Dictionary = Vec<DictionaryEntry>;
+pub struct DictionaryEntry {
+    key: String,
+    elem: Box<dyn Encodable>
 }
 
-impl Literal for Dict {
-    fn stringify(&self) -> Result<String, Error> {
-        let mut ascii = String::new();
-        for (key, value) in self.content.iter() {
-            let key = key.stringify()?;
-            let value = value.stringify()?;
-            ascii.push_str(key.as_str());
-            ascii.push_str(value.as_str());
+impl DictionaryEntry {
+    /// Create a new dictionary entry from a key and an encodable element
+    pub fn new<T: Encodable+'static>(key: &str, elem: T) -> Self {
```

Comment:
- `<T: Encodable+'static>` Why is there a `+`, and also why do we need static lifetime?

Reply:
`Encodable+'static` is used as a constraint to the generic parameter `T` .

`'static` is needed to pass `elem` into a new `Box` as `Box<dyn Encodable>` is shorthand for `Box<dyn Encodable + 'static>`. It needs to be explicitly present as a parameter.

Reply:
Ty, makes more sense now. I heard boxes allow for storing values with sizes unknown by the compiler. Does the reason behind enforcing static lifetime during allocation of generic values relate to this?

Reply:
`Box` is a pointer. I believe by requiring the content of a `Box` to have a static lifetime just makes box safe to dereference.